tfx/experimental/pipeline_testing/pipeline_recorder_utils.py

```diff
@@ -0,0 +1,151 @@
+# Lint as: python3
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Recording pipeline from MLMD metadata."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from collections import defaultdict
+import os
+
+import absl
+from ml_metadata.proto import metadata_store_pb2
+import tensorflow as tf
+from typing import Dict, List, Text
+
+from tfx.orchestration import metadata
+from tfx.utils import io_utils
+
+def get_paths(metadata_connection: metadata.Metadata,
+              executions: Dict[Text, List[metadata_store_pb2.Execution]],
+              output_dir: Text):
+  """Returns a zipped list of source artifact uris and destination uris.
+  Destination uris are stored in output_dir.
+  Args:
+    metadata_connection: A class for metadata I/O to metadata db.
+    executions:
+    output_dir: Uri to metadata db.
+
+  Returns:
+    Zip of src_uris and dest_uris.
+  """
+  events = [
+      x for x in metadata_connection.store.get_events_by_execution_ids(
+          [e.id for e in executions])
+      if x.type == metadata_store_pb2.Event.OUTPUT
+  ]
+  unique_artifact_ids = list({x.artifact_id for x in events})
+
+  src_uris = []
+  dest_uris = []
+  for artifact in \
+        metadata_connection.store.get_artifacts_by_id(unique_artifact_ids):
+    src_uris.append(artifact.uri)
+    component_id = \
+        artifact.custom_properties['producer_component'].string_value
+    name = artifact.custom_properties['name'].string_value
+    dest_uris.append(os.path.join(output_dir, component_id, name))
+  return zip(src_uris, dest_uris)
```

Comment:
Suggestion: One more pythonic way would be a generator.
For example, we can `yield (src_uri, dest_uri)` in the for loop.